pkg/util/ioutil/ioutil.go

```diff
@@ -7,71 +7,59 @@ package ioutil
 import (
 	"context"
 	"io"
+	"sync/atomic"
 	"time"
 )
 
 // timeoutReader monitors the progress of the io.Reader Read method
-// and calls the cancel function when reading does not progress for
+// and calls the cancelFn function when reading does not progress for
 // the specified time.
 type timeoutReader struct {
-	t *time.Timer
-	r io.Reader
-	n chan int
+	total  atomic.Uint64 // Total number of bytes read until timeout.
+	reader io.Reader
+
+	timer   *time.Timer
+	timeout time.Duration
+
+	done   atomic.Bool
+	cancel context.CancelFunc
 }
 
 // Read implements the io.Reader Read interface.
 func (tr *timeoutReader) Read(p []byte) (int, error) {
-	n, err := tr.r.Read(p)
-
-	v := n
-	if err != nil && tr.t.Stop() {
-		select {
-		case <-tr.t.C:
-		default:
+	n, err := tr.reader.Read(p)
+	if !tr.done.Load() && tr.timer.Reset(tr.timeout) {
+		tr.total.Add(uint64(n))
+		if err != nil {
+			tr.cancel()
+			tr.done.Store(true)
 		}
-
-		// Negative value signals that error has
-		// occurred and the goroutine should terminate.
-		v = -1
 	}
-
-	select {
-	case tr.n <- v:
-	default:
-	}
-
 	return n, err
 }
 
 // TimeoutReader creates a new timeoutReader instance and starts
 // a goroutine that monitors the progress of reading from the given reader.
 // If no progress is made for the duration of the given timeout, then the
-// reader executes the given cancel function and terminates the underlying
+// reader executes the given cancelFn function and terminates the underlying
 // goroutine. This goroutine will also terminate when the given context
 // is canceled or if an error is returned by the given reader.
-func TimeoutReader(ctx context.Context, r io.Reader, timeout time.Duration, cancel func(uint64)) io.Reader {
-	tr := &timeoutReader{t: time.NewTimer(timeout), r: r, n: make(chan int)}
+func TimeoutReader(ctx context.Context, reader io.Reader, timeout time.Duration, cancelFn func(uint64)) io.Reader {
+	ctx, cancel := context.WithCancel(ctx)
+	tr := &timeoutReader{
+		reader:  reader,
+		timer:   time.NewTimer(timeout),
+		timeout: timeout,
+		cancel:  cancel,
+	}
 
 	go func() {
-		var total uint64 = 0
-
-		for {
-			select {
-			case n := <-tr.n:
-				switch {
-				case n < 0:
-					return
-				case n >= 0:
-					tr.t.Reset(timeout)
-					total += uint64(n)
-				}
-			case <-tr.t.C:
-				cancel(total)
-				return
-			case <-ctx.Done():
-				return
-			}
+		select {
+		case <-ctx.Done():
+		case <-tr.timer.C:
+			cancelFn(tr.total.Load())
 		}
+		tr.done.Store(true)
```

Comment:
would it also make sense to stop timer here? because for normal reads, when reading is executed before timeout, it would make sense to stop timer in order to gracefully clean up resources.

Reply:
Done.